cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,238 @@
+      LOG.debug("Fetch request for searchOffset={}", searchOffset);
+      ByteBufferMessageSet messageSet
+        = getResponseMessageSet(consumer, partition, clientName, BUFFER_SIZE, searchOffset);
+      for (MessageAndOffset messageAndOffset : messageSet) {
+        try {
+          long time = serializer.decodeEventTimestamp(messageAndOffset.message().payload());
+          if (time > maxTime) {
+            LOG.debug("Failed to find the message with timestamp {} in topic {} partition {} after " +
+                        "exceeding the max time limit {} with time {}. Returning the next offset {} of the message " +
+                        "with largest offset and event time before minTime=(targetTime - PIPELINE_EVENT_DELAY_MS)={}." +
+                        " Or the starting offset {} if no message has event time before minTime.",
+                      targetTime, topic, partition, maxTime, time, minTime, startOffset);
+            // Return the next offset before minTime if no message with targetTime is found
+            return nextOffsetBeforeMinTime;
+          }
+          if (time == targetTime) {
+            LOG.debug("Matched offset {} for time {} found", messageAndOffset.nextOffset(), time);
+            return messageAndOffset.nextOffset();
+          }
+          // Update nextOffsetBeforeMinTime if time < minTime
+          if (time < minTime) {
+            nextOffsetBeforeMinTime = messageAndOffset.nextOffset();
+          }
+          searchOffset = messageAndOffset.nextOffset();
+        } catch (IOException e) {
+          LOG.warn("Message with offset {} in topic {} partition {} is ignored because of failure to decode.",
+                   messageAndOffset.nextOffset(), topic, partition, e);
+        }
+      }
+    }
+    LOG.debug("Failed to find message with event time equal to targetTime={} after searching until offset {}. " +
+                "Returning the next offset {} of the message with largest offset and event time before" +
+                " minTime=(targetTime - PIPELINE_EVENT_DELAY_MS)={}. Or the starting offset {} if no message has" +
+                "event time before minTime.",
+              targetTime, searchOffset, nextOffsetBeforeMinTime, minTime, startOffset);
+    // Return nextOffsetBeforeMinTime if no message with targetTime is found or maxTime is not reached
+    return nextOffsetBeforeMinTime;
+  }
+
+  /**
+   * Fetch a log event with {@code requestOffset} and deserialize it to get the log event time.
+   *
+   * @return the log event time of the message with {@code requestOffset} or {@code -1} if no message found.
+   */
+  private long getEventTimeByOffset(SimpleConsumer consumer, int partition, String clientName, long requestOffset) {
+    ByteBufferMessageSet messageSet
+      = getResponseMessageSet(consumer, partition, clientName, SINGLE_MESSAGE_MAX_SIZE, requestOffset);
+    for (MessageAndOffset messageAndOffset : messageSet) {
+      long offset = messageAndOffset.offset();
```

Comment:
so its highly likely that you can get more than one message in messageSet right, can we check if `messageAndOffset.offset() == requestOffset` then decode that message ? rather than decoding the first message in the set.